config/cloudpersisters/google/datastore_persister.go

```diff
@@ -0,0 +1,217 @@
+// Licensed under the Apache License, Version 2.0
+// Details: https://raw.githubusercontent.com/maniksurtani/quotaservice/master/LICENSE
+
+// Package google implements a ConfigPersister making use of Google Cloud's Datastore to store configuration.
+// See https://cloud.google.com/datastore/ for more details.
+//
+// Configuration
+// -------------
+// This package expects a Google Cloud account set up, and should be initialized with a JSON file containing credentials
+// with which to connect to Google Cloud. Google Cloud Datastore should be enabled on your account. A JSON credentials
+// file can be created by browsing to https://console.cloud.google.com/iam-admin/serviceaccounts and selecting your
+// project, and creating a new key.
+//
+// The Datastore should further be configured with a namespace and an entity type, both of which should be passed in
+// to the `New()` function when creating a new DatastoreConfigPersister. Do not create any entity instances by hand;
+// they should only ever be created by this package.
+//
+// Once configurations are persisted, they can be viewed on the Google Cloud admin dashboard.
+package google
+
+import (
+	"io"
+	"cloud.google.com/go/datastore"
+	"golang.org/x/net/context"
+	"google.golang.org/api/option"
+	"github.com/maniksurtani/quotaservice/config"
+	"fmt"
+	"bytes"
+	"github.com/maniksurtani/quotaservice/logging"
+	"io/ioutil"
+	"time"
+	"strings"
+	"errors"
+	"strconv"
+)
+
+type storedEntity struct {
+	Contents []byte
+	Version int32
+	Date time.Time
+	User string
+	Hash string
+}
+
+// DatastoreConfigPersister is a config persister that makes use of Google Cloud's Datastore service.
+type DatastoreConfigPersister struct {
+	projectId string
+	namespace string
+	entity string
+	client *datastore.Client
+	*config.Notifier
+	version int
+	newVersions chan int
+}
+
+func (p *DatastoreConfigPersister) PersistAndNotify(r io.Reader) error {
+	// Persist...
+	b, e := ioutil.ReadAll(r)
+	if e != nil {
+		return e
+	}
+
+	cfg, e := config.Unmarshal(bytes.NewReader(b))
+	if e != nil {
+		return e
+	}
+
+	s := &storedEntity{Contents: b,
+		Version: cfg.Version,
+		Date: time.Unix(cfg.Date, 0),
+		User: cfg.User,
+		Hash: config.HashConfig(b)}
+
+	k := datastore.NameKey(p.entity, fmt.Sprintf("version:%v", cfg.Version), nil)
+	k.Namespace = p.namespace
+
+
+	_, e = p.client.RunInTransaction(context.Background(), func (tx *datastore.Transaction) error {
+		existing := &storedEntity{}
+		e := tx.Get(k, existing)
+		if e != nil && e != datastore.ErrNoSuchEntity {
+			return e
+		}
+
+		if e == nil {
+			// Fail; this version already exists; ignore?
```

Comment:
If e == nil we skip the put (so we don't overwrite). A `Put` would overwrite if that same version already exists.

Re: the stored entity being different, we could check for type. Not sure if we want to do a deep equals to compare contents?